### PR TITLE
An attempt to work around the DataContext grouping issue #2937

### LINKED ISSUE #2937: A simple query with grouping throws a NullReferenceException
My application has recently migrated from BLToolkit to Linq2db.
The app has lots of simple queries like this:

```csharp
using (var db = new DataContext())
{
	var dictionary = db.GetTable<PostgresTable>()
		.GroupBy(g => g.TableType, e => e.TableName)
		.ToDictionary(p => p.Key);

	var tables = dictionary.ToDictionary(p => p.Key, p => p.Value.ToList());
}

[Table(Name = "information_schema.tables")]
public class PostgresTable
{
	[Column(Name = "table_name")]
	public string TableName { get; set; }

	[Column(Name = "table_type")]
	public string TableType { get; set; }
}
```

All these queries with grouping throw a NullReferenceException when executed:

```
Exception message: Ссылка на объект не указывает на экземпляр объекта.
Stack trace:
   в LinqToDB.DataContext.QueryRunner.get_DataContext()
   в LinqToDB.Linq.Builder.GroupByBuilder.GroupByContext.Grouping`2.GetItems()
   в LinqToDB.Linq.Builder.GroupByBuilder.GroupByContext.Grouping`2.GetEnumerator()
   в System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   в System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   в Program.<>c.<Main>b__0_4(KeyValuePair`2 p)
   в System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, IEqualityComparer`1 comparer)
   в System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector)
   в Program.Main()
```

### Steps to reproduce

The full application that reproduces the issue follows:

```c#
using System;
using System.Data;
using System.Data.Linq.Mapping;
using System.Linq;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Transactions;
using LinqToDB;
using LinqToDB.Data;
using LinqToDB.DataProvider.PostgreSQL;

class Program
{
	const string ConnectionString = "Host=localhost;Port=1522;Database=postgres;User Id=test;Password=test";

	[Table(Name = "information_schema.tables")]
	public class PostgresTable
	{
		[Column(Name = "table_name")]
		public string TableName { get; set; }

		[Column(Name = "table_type")]
		public string TableType { get; set; }
	}

	static void Main()
	{
		SetupDataConnection();

		using (var db = new DataContext())
		{
			var dictionary = db.GetTable<PostgresTable>()
				.GroupBy(g => g.TableType, e => e.TableName)
				.ToDictionary(p => p.Key);

            var tables = dictionary.ToDictionary(p => p.Key, p => p.Value.ToList());
		}
	}

	static void SetupDataConnection()
	{
		// use PostgreSQL data provider
		var provider = new PostgreSQLDataProvider();
		Console.WriteLine($"Initializing the data provider: {provider.Name}...");

		DataConnection.DefaultDataProvider = provider.Name;
		DataConnection.DefaultConfiguration = "*";
		DataConnection.AddConfiguration("*", ConnectionString, provider);

		// set up tracing
		DataConnection.TurnTraceSwitchOn();
		DataConnection.WriteTraceLine = (message, displayName, level) =>
			Console.WriteLine($"MESSAGE: {message}\nDISPLAY_NAME: {displayName}");
	}
}
```

### Environment details

linq2db version: 3.3.0
Database Server: PostgreSQL v9.6, v11, v13
Database Provider: Npgsql 4.1.9
Operating system: Windows 10 64-bit
.NET Framework: 4.61

### PR DESCRIPTION
**Pull request by yallie**
Fix #2937

Hi, this pull request has two unit tests that demonstrate the NullReferenceException
when a query with a grouping is executed by `DataContext`, see issue #2937.

I tried to work around the NRE, but now I'm getting an `ObjectDisposedException` instead:

```csharp
using (var dc = new DataContext(context))
{
    // part1: this part works fine
	var dictionary = dc.GetTable<Person>()
		.GroupBy(p => p.FirstName)
		.ToDictionary(p => p.Key);

    // part2: this part throws an ObjectDisposedException 
	var tables = dictionary.ToDictionary(p => p.Key, p => p.Value.ToList());
}
```

That's because the DataContext's query runner redirects all its calls to the
internal DataConnection's query runner, and the data connection is released 
when the first part is executed and DataContext.ReleaseQuery is called.

Any thoughts how this can be addressed?

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/linq2db/linq2db?pullRequest=2999) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/linq2db/linq2db?pullRequest=2999) before we can accept your contribution.<br/><br/>:x: yallie [sign now](https://cla.dotnetfoundation.org/linq2db/linq2db?pullRequest=2999)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/linq2db/linq2db?pullRequest=2999) it.</sub>

**Comment by yallie** [comment]
The problem is that `Grouping<TKey,TElement>` takes the query runner in the constructor 
and then later tries to reuse it when the query runner is disposed by the DataContext:

[GroupByBuilder.cs#L280](https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/Linq/Builder/GroupByBuilder.cs#L280)

internal class Grouping<TKey,TElement> : IGrouping<TKey,TElement>
{
	public Grouping(
		TKey                    key,
		IQueryRunner            queryRunner,
		List<ParameterAccessor> parameters,
		Func<IDataContext,TKey,object?[]?,IQueryable<TElement>> itemReader)
	{
        ...
    }
    ...

    List<TElement> GetItems()
	{
        // this one's called when _queryRunner is closed
		using (var db = _queryRunner.DataContext.Clone(true))
		{
			...
        }
    }
}
```

Shouldn't it be done the opposite way? Like, take the data context in the constructor,
then later clone it as needed and get the QueryRunner using dataContextClone.GetQueryRunner?

**Comment by sdanyliv** [comment]
I would suggest to do grouping on the client side. SQL has limitation, you can get only grouping key and aggregation result.
```cs
var dictionary = dc.GetTable<Person>()
                .AsEnumerable()
		.ToDictionary(p => p.FirstName);
```

**Comment by yallie** [comment]
@sdanyliv, thanks for the suggestion, I'm aware of this limitation.
We already discussed it in #2937.

I have a large BLToolkit-based application with hundreds of queries written like that.
I'm porting it to Linq2db and I need to make sure that all queries work unchanged.
Unfortunately, I have no control of the queries written by end-users.

**Comment by yallie** [comment]
This pull request actually works around the issue.
Can you please review my changes?

**Comment by sdanyliv** [comment]
Yes, sure, later today.

File: Tests/Linq/Data/DataExtensionsTests.cs [hunk 1]
```diff
@@ -171,6 +172,67 @@ public void TestObjectLeftJoinProjection([IncludeDataSources(TestProvName.AllSQL
 			}
 		}
 
+		[Test]
+		public void TestGrouping1([IncludeDataSources(TestProvName.AllSQLite)] string context)
+		{
+			var old = (Configuration.Linq.GuardGrouping, Configuration.Linq.PreloadGroups);
+			Configuration.Linq.GuardGrouping = false;
```
Comment (MaceWindu) [inline Tests/Linq/Data/DataExtensionsTests.cs:179]: use already existing disposable classes, defined in TestBase for scoped options:
https://github.com/linq2db/linq2db/blob/master/Tests/Base/TestBase.cs#L1412
https://github.com/linq2db/linq2db/blob/master/Tests/Base/TestBase.cs#L1442

**Review by MaceWindu — CHANGES_REQUESTED** [review]
Please update tests a bit.
Looks correct. Actually I've already faced similar issue and made changes to DataContext here #2941 for future v4

**Review by MaceWindu — APPROVED** [review]

**Review by sdanyliv — APPROVED** [review]

**Pull request merged**